[TOPIC-GPIO] gpio: Update mcux igpio driver to use new gpio api

MaureenHelm opened the pull request:
Updates the mcux igpio driver and all associated boards to use new
device tree compatible gpio configuration flags. Implements new port
get/set/clear/toggle and pin_interrupt_configure functions recently
added to the gpio api.

Note that the driver now handles configuring by port inefficiently,
making separate register writes for each pin in the port. However, port
configuration is deprecated in the new gpio api and will soon be
removed.

Tested with:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_api_1pin

On boards:
- mimxrt1015_evk
- mimxrt1020_evk
- mimxrt1050_evk
- mimxrt1060_evk
- mimxrt1064_evk

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
gpio: Update mcux igpio driver to use new gpio api

Updates the mcux igpio driver and all associated boards to use new
device tree compatible gpio configuration flags. Implements new port
get/set/clear/toggle and pin_interrupt_configure functions recently
added to the gpio api.

Assumes the gpio api layer handles translating logical flags to physical
flags.

Removes port configuration support since that feature is deprecated in
the new gpio api.

Tested with:
- samples/basic/blinky
- samples/basic/button
- tests/drivers/gpio/gpio_api_1pin

On boards:
- mimxrt1015_evk
- mimxrt1020_evk
- mimxrt1050_evk
- mimxrt1060_evk
- mimxrt1064_evk

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: mimxrt1020_evk: Fix button label to match silkscreen

Fixes the user button label in the device tree to match the board's
silkscreen.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

agansari reviewed (approved):
Code is OKAY.
Tested on i.mx 1050 rt, works as expected.

pabigot commented:
Has anybody run the 2-pin test #19249 on this?  Hack it to specify usable pins and do pinmux if necessary.

MaureenHelm commented:
Rebased to incorporate #19248. Still need to run the 2-pin test

mnkp commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -18,57 +18,51 @@ struct mcux_igpio_config {
 	u32_t pin_callback_enables;
 };
 
+static int mcux_igpio_pin_interrupt_configure(struct device *dev,
```
Since this function is not called by `mcux_igpio_configure` anymore we could skip the declaration.

mnkp commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -18,57 +18,51 @@ struct mcux_igpio_config {
+		return -ENOTSUP;
+	}
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
```
As stated in the PR's description access by port is not very efficient and soon will be removed from the driver code. I propose to return here -ENOTSUP rather than attempt to configure all the pins. It should be easier to clean up later the driver code. This branch should not be executed anyway since access by port is already removed from the API.

mnkp commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -18,57 +18,51 @@ struct mcux_igpio_config {
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
```
It would be nice if the configure function returned `-ENOTSUP` for `GPIO_PULL_UP`, `GPIO_PULL_DOWN` flags. The feature is currently not verified by the testcase but we should soon add the check.

mnkp commented on drivers/gpio/gpio_mcux_igpio.c:
```diff
@@ -102,6 +96,98 @@ static int mcux_igpio_read(struct device *dev,
+	} else if (pin < 32) {
+		shift = 2 * (pin - 16);
+		base->ICR2 = (base->ICR2 & ~(3 << shift)) | (icr << shift);
+	} else {
```
There is an assert in the API that guarantees that pin number is smaller than the maximum port width. Since the maximum port width in case of mcux driver is 32 we could skip the `else`.

galak commented:
Need to be rebased on updated topic-gpio branch.

MaureenHelm committed:
tests: gpio: Enable gpio_basic_api test on mimxrt1050_evk

Enables the 2-pin gpio test on the mimxrt1050_evk board by adding a dts
overlay and configuring pinmuxes on the arduino header.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
Rebased to pick up 2-pin test, which now passes. The problem I had earlier was fixed by locking interrupts in `mcux_igpio_pin_interrupt_configure`

agansari commented on tests/drivers/gpio/gpio_basic_api/boards/mimxrt1050_evk.overlay:
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
change copyright

MaureenHelm replied:
I kept the Nordic copyright because I copied the file from another board

agansari commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -34,6 +38,22 @@ static void board_setup(void)
+			    IOMUXC_SW_PAD_CTL_PAD_PKE_MASK |
+			    IOMUXC_SW_PAD_CTL_PAD_SPEED(2) |
+			    IOMUXC_SW_PAD_CTL_PAD_DSE(6));
+#endif
```
i moved this to board's pinmux initialization

MaureenHelm replied:
Yes this is ugly here, but I also don't want to make it the default pinmux configuration for the board. I think it's more likely these pins will be used for UART.

carlescufi commented:
@MaureenHelm 
The topic branch has been rebased. Could you please rebase this PR against it?

MaureenHelm commented:
> The topic branch has been rebased. Could you please rebase this PR against it?

Done
